### DISCUSSION TITLE
TypeError: 'int' object is not subscriptable

### DISCUSSION THREAD

**Discussion by SnMtlb** [dipy/dipy#2607] [Tractography / Tractometry]
I get this error while running "response, ratio = auto_response_ssst(gtab, data, roi_radii=10, fa_thr=0.7)" for HARDI deterministic in diffusion MRI analysis
how can I fix it?

**Comment by skoudoro** [comment]
Hi @SnMtlb,

we need the full error and a part of the code here. We can not answer this question with so little information. it could be everything.

**Reply by SnMtlb** [reply]
Dear @skoudoro 

**the code is:::**

from dipy.reconst.csdeconv import auto_response_ssst
import numpy as np

response, ratio = auto_response_ssst(gtab, data, roi_radii=10, fa_thr=0.5)

np.savetxt(os.path.join(out_dir, 'frf.txt'), np.hstack([response[0], response[1]]))



**the error is long as below:**
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-9e80c7164d6c> in <module>
      2 import numpy as np
      3 
----> 4 response, ratio = auto_response_ssst(gtab, data, roi_radii=10, fa_thr=0.5)
      5 
      6 # Save the FRF

~/.local/lib/python3.8/site-packages/dipy/reconst/csdeconv.py in auto_response_ssst(gtab, data, roi_center, roi_radii, fa_thr)
   1055 
   1056     mask = mask_for_response_ssst(gtab, data, roi_center, roi_radii, fa_thr)
-> 1057     response, ratio = response_from_mask_ssst(gtab, data, mask)
   1058 
   1059     return response, ratio

~/.local/lib/python3.8/site-packages/dipy/reconst/csdeconv.py in response_from_mask_ssst(gtab, data, mask)
   1014     S0s = data[indices][:, np.nonzero(gtab.b0s_mask)[0]]
   1015 
-> 1016     return _get_response(S0s, lambdas)
   1017 
   1018 

~/.local/lib/python3.8/site-packages/dipy/reconst/csdeconv.py in _get_response(S0s, lambdas)
   1068         return response, ratio
   1069     l01 = np.mean(lambdas, axis=0) if S0s.size else 0
-> 1070     evals = np.array([l01[0], l01[1], l01[1]])
   1071     response = (evals, S0)
   1072     ratio = evals[1] / evals[0]

TypeError: 'int' object is not subscriptable

**Comment by arokem** [comment]
I think this might be a bug. These lines of code: https://github.com/dipy/dipy/blob/master/dipy/reconst/csdeconv.py#L1069-L1070, don't make sense to me, because if `l01 = 0` this error will always be raised when indexing on the second line of code.

**Reply by arokem** [reply]
@SnMtlb : could you please tell us a bit more about this data? What is the dimensionality of the `data` variable? What are the b-values in `gtab`?

**Reply by SnMtlb** [reply]
Dear @arokem,
I use a subject from HCP preprocessed diffusion data.

The **shape** of variable data is (173, 207, 173, 143)
The shape of variable bvals is (143,)
The shape of variable bvecs is (143, 3)

when I run "**gtab = gradient_table(bvals, bvecs)**", I get this error:
/home/sana/.local/lib/python3.8/site-packages/dipy/core/gradients.py:300: UserWarning: b0_threshold (value: 50) is too low, increase your b0_threshold. It should be higher than the lowest b0 value (55.0).
  warn("b0_threshold (value: {0}) is too low, increase your "

**gtab.b0s_mask** is:
[False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False]


143th experiment was acquired with b-value=2020.0 s/mm^2
143th experiment was acquired with b-vector=[0.459878 0.407046 0.789193]
b-value = [  65.  980. 1980.  990. 2020. 1010. 1985. 1000. 2025.  975.   60. 2000.
  990. 1990. 1010. 1980. 1015. 2020.  990. 2015.   55.  995. 2005.  995.
 2005.  980. 1970. 1015. 1995.  995.   65. 2020.  990. 2015. 1015. 2010.
  985. 1965.  985. 2005.   60. 1005. 1990. 1015. 1980.  990. 2020.  995.
 2010.  980.   55. 2020. 1000. 2000. 1020. 1995.  990. 1985.  985. 2015.
   60.  995. 2015. 1015. 1970. 1005. 1995. 1020. 2015. 1000.   65. 2015.
 1010. 2020. 1005. 2025.  980. 1995.  985. 2000.   65. 1005. 2010.  985.
 1975. 1010. 1975.  985. 1995.  990.   60. 1990.  975. 1965. 1005. 1995.
  995. 2020. 1010. 1985.   60. 1015. 2015.  995. 1975. 1000. 2005.  985.
 2000.  980.   60. 1975.  985. 2025. 1005. 1985. 1010. 1990.  980. 1985.
   65. 1005. 1970. 1005. 2005. 1020. 1985. 1000. 1985. 1010.   65. 1970.
  980. 2015.  985. 1975. 1000. 2010. 1000. 1985.   60.  985. 2020.] s/mm^2

Regards,

**Reply by arokem** [reply]
Yeah. I think that the problem is that you don't have any b0s identified in the gradient table and that triggers the bug I mentioned.

**Reply by SnMtlb** [reply]
> Yeah. I think that the problem is that you don't have any b0s identified in the gradient table and that triggers the bug I mentioned.

But, most of the 7T data don't have any b0. Does it mean that I can not use this code for 7T data? or is there any exchangeable way to use the same code?

**Reply by arokem** [reply]
You could designate those low b-value measurements as b0s by setting `b0_threshold=65` when initializing the gradient table object. Otherwise, you'll hit this bug (again, I think the code is not working as intended, and this shouldn't be a problem... but it is)

**Reply by SnMtlb** [reply]
Thank you so much for your help. It is really appreciated.
